src/Microsoft.ML.Transforms/Text/NgramHashingTransformer.cs

```diff
@@ -288,506 +244,456 @@ public ColInfoEx(ModelLoadContext ctx)
+        {
+            _columns = columns.ToImmutableArray();
+            foreach (var column in _columns)
+            {
+                if (column.InvertHash != 0)
+                    throw Host.ExceptParam(nameof(columns), $"Found colunm with {nameof(column.InvertHash)} set to non zero value, please use { nameof(NgramHashingEstimator)} instead");
+            }
+        }
 
-        /// <summary>
-        /// Public constructor corresponding to SignatureDataTransform.
-        /// </summary>
-        public NgramHashingTransformer(IHostEnvironment env, Arguments args, IDataView input)
-            : base(env, RegistrationName, input)
+        internal NgramHashingTransformer(IHostEnvironment env, IDataView input, params ColumnInfo[] columns) :
+            base(Contracts.CheckRef(env, nameof(env)).Register(nameof(NgramHashingTransformer)))
         {
-            Host.CheckValue(args, nameof(args));
-            Host.CheckUserArg(Utils.Size(args.Column) > 0, nameof(args.Column));
-
-            _bindings = new Bindings(args, Source.Schema, this);
-            _exes = new ColInfoEx[args.Column.Length];
-            List<int> invertIinfos = null;
-            int[] invertHashMaxCounts = new int[args.Column.Length];
-            for (int iinfo = 0; iinfo < _exes.Length; iinfo++)
+            Contracts.CheckValue(columns, nameof(columns));
+            _columns = columns.ToImmutableArray();
+
+            int[] invertHashMaxCounts = new int[_columns.Length];
+            HashSet<int> columnWithInvertHash = new HashSet<int>();
+            HashSet<int> sourceColumnsForInvertHash = new HashSet<int>();
+            for (int i = 0; i < _columns.Length; i++)
             {
-                _exes[iinfo] = new ColInfoEx(args.Column[iinfo], args);
-                var invertHashMaxCount = GetAndVerifyInvertHashMaxCount(args, args.Column[iinfo], _exes[iinfo]);
+                int invertHashMaxCount;
+                if (_columns[i].InvertHash == -1)
+                    invertHashMaxCount = int.MaxValue;
+                else
+                    invertHashMaxCount = _columns[i].InvertHash;
                 if (invertHashMaxCount > 0)
                 {
-                    Utils.Add(ref invertIinfos, iinfo);
-                    invertHashMaxCounts[iinfo] = invertHashMaxCount;
+                    columnWithInvertHash.Add(i);
+                    invertHashMaxCounts[i] = invertHashMaxCount;
+                    for (int j = 0; j < _columns[i].Inputs.Length; j++)
+                    {
+                        if (!input.Schema.TryGetColumnIndex(_columns[i].Inputs[j], out int srcCol))
+                            throw Host.ExceptSchemaMismatch(nameof(input), "input", _columns[i].Inputs[j]);
+                        sourceColumnsForInvertHash.Add(srcCol);
+                    }
                 }
             }
 
-            InitColumnTypes();
-
-            if (Utils.Size(invertIinfos) > 0)
+            if (Utils.Size(columnWithInvertHash) > 0)
```

Comment:
Can you comment what this is doing? Also would it help to move to a Train function to separate this logic out from the constructor? Also you are evaluating the input on construction - Im assuming that is OK? That seems to go against the lazy evaluation - but that seems to be OK when we having training.

Reply:
If possible I think it would make it clearer with a separate training function.

---
In reply to: [238487172](https://github.com/dotnet/machinelearning/pull/1811#discussion_r238487172) [](ancestors = 238487172)